### DISCUSSION TITLE
Typical deployment times on HL2

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#228] [General]
In https://github.com/maluoi/StereoKit/issues/225#issuecomment-975084370 it says:

>  I think you'll be pleased at how fast SK is already! One big thing I focus on, is ensuring StereoKit builds fast enough that deploying to device takes very little time.

I was wondering if the deployment times I'm seeing are typical (and considered "very fast"). For my very simple application, already fully built (takes around 14s for a clean release-mode build), deployment to a HL2 takes at least 20 seconds if there are no changes from the last deployment. Usually, it takes a few seconds more. This is on a pretty new laptop with a Core i7 @ 2.6 GHz. This is a lot better than the multi-minute times it takes with Unity, but it's still quite some time to wait ;-)

For comparison, a clean *debug* build in x64 mode takes 16 seconds, with around 5 seconds to start in simulator mode. So that's still preferable for lowering testing iterations when I can. I haven't been able to get a debug build on ARM64 on the HL2 to start, as VS stops right after deployment is done, but maybe that's expected?

**Comment by maluoi** [comment]
An empty native UWP application takes 5-10 seconds, and .Net Native on the C# side adds a bit extra on top of this. For me at least, coming from most game engines, this is life-changingly fast!

I'd certainly agree that 20-40 seconds is not at all ideal, but I would say that StereoKit is pretty close to the limit of what is possible with the UWP platform! This is why I recommend developers use .Net Core whenever possible, and UWP only when strictly necessary. .Net Core is _much_ faster. With my own projects, I always set up a dual UWP/.Net Core solution, so I can work fast on PC, and still deploy to HL2.

I have encountered issues where UWP apps occasionally just don't start. I've generally found that uninstalling the app and running again fixes this, but I haven't figured out why? I'm not yet convinced this is an issue with StereoKit, but it is an incredibly difficult problem to inspect, so it's tough to say with certainty!

**Comment by paulmelis** [comment]
> An empty native UWP application takes 5-10 seconds, and .Net Native on the C# side adds a bit extra on top of this. For me at least, coming from most game engines, this is life-changingly fast!
> 
> I'd certainly agree that 20-40 seconds is not at all ideal, but I would say that StereoKit is pretty close to the limit of what is possible with the UWP platform

Oh, I fully agree it these are much better than using e.g. Unity and the IL2CPP workflow (although the remoting works around the slow deployment for most cases). But I find using StereoKit much more fun than using a game engine, and I've only been using it for a few days. Sure, having prior experience with using low-level scene graph librariesand writing your own game loop helps, but using SK is such a lean-and-mean way to create XR applications. For many (simpler) use cases and, for example, teaching it is a great addition!

> This is why I recommend developers use .Net Core whenever possible, and UWP only when strictly necessary. .Net Core is much faster. With my own projects, I always set up a dual UWP/.Net Core solution, so I can work fast on PC, and still deploy to HL2.

Hmm, is it possible to retrofit an existing UWP project with a .Net Core sibling, to take advantage of faster PC builds?

**Reply by maluoi** [reply]
I often start with .Net Core and add UWP later, but it's essentially the same process in reverse! You can see an example of this over in [StereoKit Ink](https://github.com/maluoi/StereoKit-PaintTutorial/tree/master/StereoKitInk_UWP), where I add .cs files from the parent directory. This works pretty well, but requires a little bit of comfort in editing .csproj files by hand, and may require some attention as your project changes structure.

I recently discovered the C# "Shared Project", and it looks very promising as a more "proper" way of doing the same thing. I don't have much experience with it, but I may be leaning into this method at some point. It may not work quite as nicely with Assets though, so fair warning there.
